### ISSUE TITLE
Administer groups and their permissions

### ISSUE THREAD

**Issue by tomcrane** [dlcs/elucidate-server#81]
From @workergnome:

> We would like an interface that allows the creator of an annotation to assign groups to an annotation at the platform level, and we would like an API to set the permission status and the groups with permissions. (Eventually, we'd like to be able to construct groups on the fly via an API and interface, but that is outside the scope of this work.)

**Comment by tomcrane** [comment]
@danielgrant  says:

> I’d much rather we exposed a service API in Elucidate, and then implemented a user interface as a separate web application which interacted with that API in the backend. This way we can keep all the logic contained within Elucidate, instead of having to implement such logic separately. 
 
> One thing to consider here is that any changes we make to implement this must not break existing W3C / OA compliance. Which means we can’t change the structure of the annotations, or modify how the mechanisms through which they are managed. I don’t think this would be an issue, but something worth keeping in mind.
 
> Also, these enhancements should be optional and, assuming Getty don’t have any objections, they should be implemented and maintained in the dlcs GitHub repository (as opposed to a fork in Getty’s GitHub). And they should be able to be turned off and on so other users of Elucidate can choose to use them or not.

**Comment by workergnome** [comment]
I think the service model is a really good one.  For the initial scope, defining the service is probably sufficient--I think a UI may be outside what's accomplishable given our initial budget, and we could work on defining that at a later date, or working on it internally.

**Comment by workergnome** [comment]
I also think that any work done to implement this should, if you're amenable, become part of the public dlcs repository.  We are interested in the functionality, but we're just as interested in helping others with the same requirements.

**Comment by workergnome** [comment]
I believe our inclusion of groups is because we know that will be a request for that feature, and we want to make sure (as I'm sure you do, too) that the implementation doesn't preclude that.  I don't think that we know yet what the best way to implement that might be--the idea we came up with were a list of explicit roles passed in at creation, pulled from a predetermined list, and to have the JWTs include those roles:

> I create a annotation that is available to all people with role "Staff".  Rob requests annotations from Elucidate with a JWT that grants him the role "Staff", and thus he can see that annotation. The list of valid roles is provided to Elucidate via configuration.

But there's almost certainly a potential feature for "I would like to share this annotation with Rob", and while that seems to be larger in scope that what we're talking about here, if we think that's a direction that Elucidate should go, how do we make sure that we don't end up with multiple paths for determining permissions--or if we do, that the design supports that.

**Comment by tomcrane** [comment]
One question - and this is almost certainly way out of scope right now - should we be looking at the hypothes.is API as a pattern for doing this? Or NOT doing like that?

https://hypothes.is/api/

**Comment by workergnome** [comment]
I have no strong feelings either way.  If you think there's benefit to consistency with it, great!  If you think that there's a better way--great!

**Comment by tomcrane** [comment]
Don't know yet! It's an anno service/server with group support, but I don't know whether its model suits yours. There may be some benefit to having Elucidate offer an API (on top of the W3C protocol) compatible with hypothes.is clients. But that is just a vague idea; it may not be possible to get any actual benefit from attempting that compatibility.

**Comment by garyttierney** [comment]
Hi @workergnome, I'm wondering if we'd be able to derive all information regarding the authentication/authorization from the token that Elucidate receives or if we need to come up with metadata to store per-group.  Would the group model be hierarchical or simply collections of users?  i.e., would staff be granted the single "staff" grant and have some groups under them that they would implicitly be able to access like so:
```
            +----------+
            |          |
            | Staff    |
            +----+-----+
                 |
                 |
        +--------+---------+
+-------v------+  +--------v------+
|  Volunteers  |  |  Other users  |
|              |  |               |
+--------------+  +---------------+
```

Or would it be more of a flat list model where the staff user would have `staff`, `volunteer`, and `other_user` grants if they needed to access annotations from the groups "below" them?

**Comment by workergnome** [comment]
I can imagine three "types" of auth grants.

1. Users have permissions for the annotations they create.

2. Users have roles included as a grant within the JWT. These roles are known ahead of time, and are static--elucidate is provided with a list of possible values, and if JWT includes the `role=staff` grant, knows what to do with that.  I would think that if someone was `staff` and also `librarian`, they would have both grants--the data would be passed within the token.  

3. Groups are lists of users constructed dynamically via API, with the collection of users stored as  metadata within elucidate somehow.  (This seems more like the hypothes.is model, from what I could glean from the quick review of their API.)

Anything else gets us deep into the authorization weeds, with Active Directory/Shiboleth/SAML/OAUTH2 discussions.  I'd love to have those discussions, but I can't imagine anything we come up with out of that being actually implementable at this budget.

**Comment by workergnome** [comment]
(which is to say, @garyttierney, that I think at the simplest level, the only thing we'd need is to know what the potential grants are.  I don't think we'd want to do anything more complicated with it--it's a list, users can have any number of grants, and it's a straight "if you have the grant, you have the permission" model, with no hierarchy)

**Comment by danielgrant** [comment]
Hi @workergnome,

What do you envisage as being the exhaustive list of roles? I would imagine that for an initial implementation, two would make sense:

- `admin`: can view and manage all annotations all of the time.
- `user`: can only view and manage annotations they created (i.e. annotations they "own"), with the exception of annotations they have access to by virtue of their group membership (subject to update / delete constraints as per https://github.com/dlcs/elucidate-server/issues/80#issuecomment-388378774).

**Comment by workergnome** [comment]
Those two seem useful as special cases, and I think I'm thinking of a third type, for now:

We pass in, via configuration, a list of strings that Elucidate considers valid roles.  There _may_ be an endpoint by which a client can enumerate those roles, but that's probably a longer discussion around security.  If a user's token has a predefined grant of one of those roles, we can consider them to have read access to any annotation that has been assigned to that role.

I would not consider it possible to define, a priori, the list of valid roles--I'm thinking of these as being, say, department-specific, and a way to control which people can see which information, not about what rights a given person has.  `librarian`, `staff` or `conservator`, not `admin` or `user`.

**Comment by danielgrant** [comment]
I worry that the sort of approach you're describing will begin to blur the line between a "group" and a "role".

We should be able to assign access to annotation(s) through group membership, but not through role assignment (with the exception of a superuser role that can view everything).

Therefore I would say `admin`, `user`, etc should be a role, and these roles should expose specific permissions - but a permission *does not* relate to an individual annotation.

In the case of `librarian`, `staff`, `conservator`, etc, these could be modelled better as groups if the intention is to expose a subset of annotations created in the system to these type of users.

**Comment by danielgrant** [comment]
Hi @workergnome,

This should summarise the enhancement:

* A new API will created and exposed by Elucidate which will allow for the management of groups and the assignment of annotations into those groups.
* Any valid authenticated user (see #80) will have the ability to create groups. The user who creates a group will be the "Owner".
* A group can only be managed by its owner or by a user with the role of `admin`.
* Group management will include:
   * Creation and deletion of groups
   * Addition and removal of users from a group
   * Addition and removal of an annotation from a group
* The API will also expose, for a given group:
   * The ID's of all annotations within that group.
   * The usernames of all users within that group.

**Comment by danielgrant** [comment]
Implemented in 1.5.0.
